### PR TITLE
Do not double-encode HTML in i18n

### PR DESCRIPTION
**Pull request by rugk**
This issue got introduced in 4bf7f86 due to double

Fixes https://github.com/PrivateBin/PrivateBin/issues/557
Fixes https://github.com/PrivateBin/PrivateBin/issues/558

Also _inverted_ the logic/variable name for containsNoLinks to
the more logical one "containsLinks" to avoid too many negations.

Also verified that the attachment name is stil properly displayed
when you clone a paste.

## ToDo

* [x] @elrido to double-check, please test the previous [exploit steps as described in our advisory again](https://github.com/PrivateBin/PrivateBin/security/advisories/GHSA-8j72-p2wm-6738)
* [x] also test with a valid a tag as the _attachment name_, i.e. `<a href="https://example.com">Test</a>`, so we can test whether an attacker could place a bad link there.

### CODE REVIEW

File: js/privatebin.js [hunk 1]
```diff
@@ -618,21 +618,15 @@ jQuery.PrivateBin = (function($, RawDeflate) {
                 args[0] = translations[messageId];
             }
 
-            // messageID may contain links, but should be from a trusted source (code or translation JSON files)
-            let containsNoLinks = args[0].indexOf('<a') === -1;
-            for (let i = 0; i < args.length; ++i) {
-                // parameters (i > 0) may never contain HTML as they may come from untrusted parties
-                if (i > 0 || containsNoLinks) {
```
Comment (rugk) [inline js/privatebin.js]: Okay, got the issue here now:
If the first part is a link, _and_ you get >1 parameter, all these are injected and applied via HTML again.

Again this leads to untrusted input in the HTML.

Here is a diff to test when that case applies that would just disallow such combinations:
```diff
diff --git a/js/privatebin.js b/js/privatebin.js
index c83d8ff..5c0aecc 100644
--- a/js/privatebin.js
+++ b/js/privatebin.js
@@ -620,6 +620,9 @@ jQuery.PrivateBin = (function($, RawDeflate) {
 
             // messageID may contain links, but only the first parameter, as that is from a trusted source (code or translation JSON files)
             let containsLinks = args[0].indexOf('<a') !== -1;
+            if (containsLinks && args.length > 0) {
+                throw new Error('security violation detected: do not concatenate links and untrusted data!');
+            }
 
             // format string
             let output = Helper.sprintf.apply(this, args);
diff --git a/tpl/bootstrap.php b/tpl/bootstrap.php
index 59c730e..0e949da 100644
--- a/tpl/bootstrap.php
+++ b/tpl/bootstrap.php
@@ -72,7 +72,7 @@ endif;
 ?>
                <script type="text/javascript" data-cfasync="false" src="js/purify-2.0.7.js" integrity="sha512-XjNEK1xwh7SJ/7FouwV4VZcGW9cMySL3SwNpXgrURLBcXXQYtZdqhGoNdEwx9vwLvFjUGDQVNgpOrT>
                <script type="text/javascript" data-cfasync="false" src="js/legacy.js?<?php echo rawurlencode($VERSION); ?>" integrity="sha512-LYos+qXHIRqFf5ZPNphvtTB0cgzHUizu2wwcOwcwz/VIpR>
-               <script type="text/javascript" data-cfasync="false" src="js/privatebin.js?<?php echo rawurlencode($VERSION); ?>" integrity="sha512-/d5+ju4pY94Jnb3o2poZNQRxKxJqgmbV4SojFKq/4G>
+               <script type="text/javascript" data-cfasync="false" src="js/privatebin.js?<?php echo rawurlencode($VERSION); ?>" integrity="sha512-NQ2iciSmIAeXCmdpJm5BGKvyGjkvN0Ob82mmbxSpgB>
                <link rel="apple-touch-icon" href="img/apple-touch-icon.png?<?php echo rawurlencode($VERSION); ?>" sizes="180x180" />
                <link rel="icon" type="image/png" href="img/favicon-32x32.png?<?php echo rawurlencode($VERSION); ?>" sizes="32x32" />
                <link rel="icon" type="image/png" href="img/favicon-16x16.png?<?php echo rawurlencode($VERSION); ?>" sizes="16x16" />
diff --git a/tpl/page.php b/tpl/page.php
index 0f8df18..ff2d5f5 100644
--- a/tpl/page.php
+++ b/tpl/page.php
@@ -50,7 +50,7 @@ endif;
 ?>
                <script type="text/javascript" data-cfasync="false" src="js/purify-2.0.7.js" integrity="sha512-XjNEK1xwh7SJ/7FouwV4VZcGW9cMySL3SwNpXgrURLBcXXQYtZdqhGoNdEwx9vwLvFjUGDQVNgpOrT>
                <script type="text/javascript" data-cfasync="false" src="js/legacy.js?<?php echo rawurlencode($VERSION); ?>" integrity="sha512-LYos+qXHIRqFf5ZPNphvtTB0cgzHUizu2wwcOwcwz/VIpR>
-               <script type="text/javascript" data-cfasync="false" src="js/privatebin.js?<?php echo rawurlencode($VERSION); ?>" integrity="sha512-/d5+ju4pY94Jnb3o2poZNQRxKxJqgmbV4SojFKq/4G>
+               <script type="text/javascript" data-cfasync="false" src="js/privatebin.js?<?php echo rawurlencode($VERSION); ?>" integrity="sha512-NQ2iciSmIAeXCmdpJm5BGKvyGjkvN0Ob82mmbxSpgB>
                <link rel="apple-touch-icon" href="img/apple-touch-icon.png?<?php echo rawurlencode($VERSION); ?>" sizes="180x180" />
                <link rel="icon" type="image/png" href="img/favicon-32x32.png?<?php echo rawurlencode($VERSION); ?>" sizes="32x32" />
                <link rel="icon" type="image/png" href="img/favicon-16x16.png?<?php echo rawurlencode($VERSION); ?>" sizes="16x16" />
```

**Comment by rugk** [comment]
The problem is the vulnerable case triggers quite often, e.g. for `Your paste is <a id=\"pasteurl\" href=\"%s\">%s</a> <span id=\"copyhint\">(Hit [Ctrl]+[c] to copy)</span>`.
In this case, the input is however completely trusted, which is awkward.

So actually, we _might_ be safe with this PR as long as there are no cases where:
* both a link is used as first param
* _and_ you get untrusted (really untrusted!) input

In such a case, the attack is also mitigated by DOMPurify, but an attacker could e.g.
* inject line breaks, links etc., i.e. all whitelisted attributes
* the previously simple attack with "<script" tags or so is not possible anymore in any case

So what can we do?
* we could manually go through all code paths that have encoding issues; introduce another parameter to mark an I18n string as "trusted" and _only_ then allow HTML. This would ensure we check the risk of each of the untrusted input parameters, i.e. check whether they are really untrusted.
* or do something more sophisticated basically faking sprintf and injecting the untrusted parameters via another textNode or so, or. This would be hard.
* ~~Or we try some middle ground by not escaping _all_ characters that may be used in the parameters later, but uhm…~~ forget it, the characters that are escaped are already not _that_

---
also uhm, why do you use some textarea there… what??? for escaping?

Can we please bring back [the old mapping function](https://github.com/PrivateBin/PrivateBin/commit/4bf7f863dc2ffea1ea105e575205ab0f83ed2751#diff-fde4e313d5f4400cd7405ed9baa7053cL46)??

Maybe it is some, but is is awkward. If we have a string, just replace all chars and that's it?? (as done before)

**Comment by rugk** [comment]
Also the old mapping function escaped more chars, so could we use this as a "untrusted escape" thing and maybe then have a case for "trusted" parameters, where we can apply the "not-so-hard" escaping. Or well… none…

**Comment by rugk** [comment]
Okay the unit tests are fun, because they now already show the cases, we need to handle… Some… :smile: 

Anyway, what is your idea to move forward, @elrido?

The current state of this PR is: would keep security issues fixed and fixes the encoding issues, but breaks the whole application in some insecure cases

We now "only" need to fix the last thing. :smile:

File: js/privatebin.js [hunk 2]
```diff
@@ -618,21 +618,18 @@ jQuery.PrivateBin = (function($, RawDeflate) {
...
```
Comment (rugk) [inline js/privatebin.js]: Wait, is not this the bug actually? `i > 0` includes the first parameter, too, does not it??

**Comment by rugk** [comment]
BTW FYI https://cheatsheetseries.owasp.org/cheatsheets/DOM_based_XSS_Prevention_Cheat_Sheet.html is the guide we have to look at, but hmm… hard thing.

**Comment by elrido** [comment]
Reg. the old mapping function: You had always been wary of the use of regex in it and we had in all instances that still needed it, switched to the jQuery based substitution of `$('<div>').text(the_string_to_encode).html()`. But I had found some references stating that the use of a div leads to issues by itself, as code injected in that node may still allow risky scripts to be executed in the detached div, at least in older jQuery versions:
- https://stackoverflow.com/questions/1219860/html-encoding-lost-when-attribute-read-from-input-field/1219983#1219983
- https://stackoverflow.com/questions/1912501/unescape-html-entities-in-javascript/1912522#1912522

I had therefore used the textarea and a version that doesn't use jQuery to re-implement this. The additional characters slipping through I found in unit testing before committing (something that I highly recommend you to do and it is [very easy to do so now with the janitor or unit testing container images](https://github.com/PrivateBin/PrivateBin/wiki/Docker#unit-test-suite-privatebinunit-testing), so no excuses for not doing it, please).

On re-reading the above links, I notice that both replies got edited to [suggest the DOMParser instead](https://stackoverflow.com/questions/1912501/unescape-html-entities-in-javascript/34064434#34064434). JSdom supports this as well, it just [has to be exposed](https://stackoverflow.com/questions/42649700/using-domparser-in-javascript-testing-with-mocha-and-jsdom/47435497#47435497) in common.js, in our case.

**Comment by rugk** [comment]
All your links are about _unescaping_ trusted input or so. That's not what we want, we want to escape it. 🤔 

And [as far as I see](https://github.com/PrivateBin/PrivateBin/commit/4bf7f863dc2ffea1ea105e575205ab0f83ed2751#diff-fde4e313d5f4400cd7405ed9baa7053cL110) the old method just used string replacement, no div, no textarea, not interacting with any element in the DOM. 🤔

**Comment by elrido** [comment]
@rugk: Please read the links more closely, one of them give both en- and decoding examples using textarea, the other is provided as an indication why textarea is to be prefered over div.

The old method was regex based, but you didn't like that approach, so we had all of it's uses removed and replaced by the div based jQuery hack, see a0740ff79f9076ec7fa4d80bdfb32337a7136482.

In any case, I was happy with the regex approach and you can certainly re-add it. Please also consider the other points I sent in the review.

**Comment by rugk** [comment]
I don't mean commit https://github.com/PrivateBin/PrivateBin/commit/a0740ff79f9076ec7fa4d80bdfb32337a7136482 from 2017, https://github.com/PrivateBin/PrivateBin/commit/4bf7f863dc2ffea1ea105e575205ab0f83ed2751 is what I talk about. It (also) removes `exports.htmlEntities`? Ah okay, that is only for a test, …

https://github.com/PrivateBin/PrivateBin/commit/a0740ff79f9076ec7fa4d80bdfb32337a7136482 replaced the htmlEntities with DomPurify – this is maybe what I wanted – and that is a good thing and the proper thing to do, yes.
Just that we now have the problem that we concatenate untrusted and trusted input…

I never said I don't like RegEx there (at least for simple replacements, maybe I did not link the link2url for obvious reasons, but that is another part). I see nothing wrong [with _that_ function](https://github.com/PrivateBin/PrivateBin/commit/a0740ff79f9076ec7fa4d80bdfb32337a7136482#diff-cf28a0bc390eee2482509aca8212f831L281-L286).

**Comment by rugk** [comment]
> My replies aren't instantly added to the PR anymore - second time that they seem to be stuck in github somehow.

AFAIK this was always the case. You need to submit your whole review when finished. (or click "add single comment", but that is ugly)

File: js/privatebin.js [hunk 3]
```diff
@@ -618,21 +618,15 @@ jQuery.PrivateBin = (function($, RawDeflate) {
...
+            // messageID may contain links, but only the first parameter, as that is from a trusted source (code or translation JSON files)
+            let containsLinks = args[0].indexOf('<a') !== -1;
 
             // format string
             let output = Helper.sprintf.apply(this, args);
 
             // if $element is given, apply text to element
             if ($element !== null) {
-                if (containsNoLinks) {
+                if (!containsLinks) {
```
[the remark this answers is not in the document]
Comment (rugk) [reply js/privatebin.js]: good idea, done

**Comment by elrido** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- js/test/I18n.js  13
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PrivateBin/PrivateBin/pullRequest?prid=4811939&bid=16005030)

File: js/privatebin.js (hunk 1 above)
Comment (elrido) [reply js/privatebin.js]: I think this should now have been addressed in 685c354d0eaa48d2871ed723d0749244e60cb4a7 - let me know over there what you think

**Comment by elrido** [comment]
Ok, so, reverted some of the changes and then, as per our discussions changed the following things:
- reverted HTML entities function to the regex based one from 2017, correcting the link to point to the current OWASP documentation that explains why all of these characters are recommended to get encoded
- while adding tests to ensure all paths in the logic get tested every time, I discovered there are actually 4 cases, now checked on every test run: output to string or into element vs first param contains link or not
- cleaned up logic as per discussion:
    - Now skips the HTML entity encoding only if we can guarantee insertion via text node.
    - When we output to string, we always encode, in case the dev decides to insert it by themselves and uses `html()` instead.
    - I did find and change one such case as well (when the created paste URL gets displayed).
- DOMpurify sanitizes gopher, ws & wss links, which we previosly had tested for - these are no longer used to generate URLs.
- Just to avoid that not encoded strings would get returned and used by a dev, even after inserting them into an element via the `translate()` function, I ensured that when an element is given, the return is empty.
- Re-tested the cases in the browser, as requested in top post.

Please [review the changes](https://github.com/PrivateBin/PrivateBin/pull/560/files) and let me know what you think.

**Comment by elrido** [comment]
Yes, there is one in the four cases that still could lead to double encoding:

> When we output to string, we always encode, in case the dev decides to insert it by themselves and uses html() instead.

In form of a decision matrix, to make it more clear:

| Will not double encode:      | $element provided | text output requested |
|------------------------------|-------------------|-----------------------|
| first param contains a link  | :heavy_check_mark: | :heavy_multiplication_x: |
| all parameters are pure text | :heavy_check_mark: | :heavy_check_mark: |

I was tempted to just require an $element in every case, so that we would have the logic to properly use safe `text()` instead of encoding requiring `html()` in a single place. But we do use the text output for the email generation, for example.

Basically I did keep the above potential for double encoding issues in there as a fail safe. Just in case that my future self, or someone else, forgets about this risk and starts to use text output again - I'll rather have them encounter double encoding instead of introducing a new security issue.

How about we review all uses of `I18n._()` / `I18n.translate()`? We should try to ensure, as much as possible, that in all cases they either pass a jquery element to apply the translation into or *do* use `html()` when inserting the text output. The latter can and should be changed, if easily doable.

**Comment by rugk** [comment]
Better idea: Can't we detect that one error in double encoding and then either a) show a warning and let the error "through" _or_, if we want to be serious, throw an error, which will certainly be noticed when you try to add this feature.

Anyway, still did not quite got the decision matrix. I still see that the encoding error happens when `!containsLinks && $element !== null`.
But that should be solvable. When `$element !== null` then just don't [do that manual encoding](https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L635-L643), as inserting via `.text` is always safe.

However, in your matrix `!containsLinks && $element !== null` would be bottom left?

Ahhh, no, I have to make a mental model of this, so here the basic code paths as they are right now:

logic | `$element !== null` | `$element === null`
-- | -- | --
`containsLinks` |  DOMPurify + set via .html  |  manual encoding + DOMPurify
`!containsLinks` |  manual encoding + set via .text | manual encoding  

(I use the variable names here to not have to translate that in my brain _again_. :wink:)

Sooo…

evaluation | `$element !== null` | `$element === null`
-- | -- | --
`containsLinks` |  attacker can inject a links or other allowed tags  |  double encoding/manual encoding not required
`!containsLinks` |  double-encoding error/manual encoding not required | manual encoding  
... |  ... | security here always depends on what is done with the text afterwards

So there are, as far as I see, there are:
*  _two_ cases, where stuff gets encoding errors
* _one_ case, which is attackable, at least an attacker can inject a link – which e.g. would still be bad for a button label or so (where users click on)

---

Given all that complexity, I'd tend we re-do that thing and have a 
`translateInsecure`, which uses `.html` with all that encoding stuff.
And the default just does not allow HTML. (and uses simple `.text`)
That should make that code way less complex and _reflects_ the security problem we still face and cannot really solve: the difference between 

When we then go through all the things that need the "insecure" version there, we do then have at least an overview of what uses HTML code there. That could then be refactored later or so and we can have a look at what gets injected there.

**Comment by elrido** [comment]
Need to think a bit about this. I rather have very rare double encoding cases in the future then re-introducing the bug via an explicitly insecure function. But the simplification of the logic is a very compelling argument. Maybe we could have the cake and eat it too?

Did do the stats on the function usage, it's not all that much, actually:

- I18n.translate() isn't directly used, so we could just remove it and move the logic into the shorthand.
- I18n._() is used 18 times:
    - 4 of them are safe (using $element, so never have issues with double encode, assuming no argument after the first is ever supposed to contain HTML)
    - 4 can easily be rewritten to use $element:
        - https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L1899
        - https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L1968
        - https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L3726
        - https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L3927
    - the following are problematic uses that require return values, but have no double encode issues (they contain risky characters):
        - https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L2102
        - https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L3266
        - https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L3659
        - https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L3673
        - https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L3681
        - https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L3754
        - https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L4502
        - https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L4505
        - https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L4508
    - 1 has a problematic use that requires a return value, and could have double encoding issues in the inserted parameter:
        - https://github.com/PrivateBin/PrivateBin/blob/42130e0468f85ab0990b4f7cf27150130c26c016/js/privatebin.js#L3665

So, apart from rewriting the obvious candidates that can be rewritten, I can only find one case that might still have double encoding issues with the logic as of the version linked above.

If we would rewrite the cases that require output in some fashion like below (or create a wrapper function that does this) we could eliminate any unsafe use of the function and simplify the logic to require the presence of $element in all cases in the original function, eliminating double encoding possibilities.

```js
let $element = $('<textarea>');
I18n._($element, 'string to translate');
return $element.text();
```

I could do this some time over the weekend, if this makes sense to you, too?

**Comment by rugk** [comment]
> the following are problematic uses that require return values, but have no double encode issues (they contain no risky characters): 

"contain no risky characters" is a misjudgement. 😄 You only look at the English string, the actual translation can contain any characters in theory.

As for rewriting the obvious things that can be rewritten: Yes, that's fine.

> 4 of them are safe (using $element, so never have issues with double encode, assuming no argument after the first is ever supposed to contain HTML)

See my table above. If `$element !== null && !containsLinks` I still expect double-encoding errors.

**Comment by elrido** [comment]
> "contain no risky characters" is a misjudgement.

Let me please phrase that differently: These cases _may not_ contain any risky characters in message ID or translation. If they do, we treat them safely, by encoding them. If this causes double encoding, then this has to be fixed in the translation files. As far as I could see, we have no case where the translated ID contains characters of that class.

If you do find such cases, please list them here or push the fixes.

Which gets us to the part that I missed reg. your interpretation for this case:

> See my table above. If $element !== null && !containsLinks I still expect double-encoding errors.

In that case we do encode all parameters, as we explicitly stated that in that case none are allowed to contain HTML or risky characters. If they do, they get encoded for _safety_, see above.

In general, what is your opinion: Do you prefer not to encode, risking HTML injection or do you want to risk double encoding in the future? At present, with the above changes, I couldn't reproduce any of the reported double encoding issues. Double encoding would only re-occur, if someone in the future misuses the function or adds something incorrect to the translation files.

**Comment by rugk** [comment]
> none are allowed to contain HTML or risky characters. If they do, they get encoded for safety, see above.

But a simple "Let's" also get's encoded to "Let&#39;s", which is a double-encoding error the users then see. If that is inside of a parameter - maybe it currently is not, but it's theoretically possible. And you can't just write improper English or so, because of that.

> Do you prefer not to encode, risking HTML injection or do you want to risk double encoding in the future?

If we ahve no choice, of course double encoding. That is at least safe.

I think the main reason why double-encoding errors do not happen currently, is because we rarely use _parameters_ for useful text.
We rather use it for links, filenames or so, i.e. arbitrary "content". Though links could also contain special characters sometimes, especially links.
I guess if we continue to avoid too many parameters, we're fine.

**Comment by elrido** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- js/test/I18n.js  13
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PrivateBin/PrivateBin/pullRequest?prid=4811939&bid=16005030)

**Comment by elrido** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- js/test/I18n.js  13
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PrivateBin/PrivateBin/pullRequest?prid=4811939&bid=16005030)

**Comment by elrido** [comment]
I updated the simple cases (aa3f120) and attempted the logic simplification (split translate into one that enforces jQuery and a wrapper that returns a string) in 6236588. Unfortunately the unit tests showed that this naive attempt would always return unsafe decoded strings, so it would just re-introduce the XSS issues - hence I reverted this again.

I did re-add the sprintf simplification and the re-phrasing of the translation jsdoc block, afterwards, in 2d11d7b.

In the current state we don't get the reported cases of double encoding anymore and yes, parameters should not be used to cobble together sentences, just to insert i.e. numbers or generated/user defined strings like the filenames, but in a secure manner. While as programmer it is tempting to build sentences out of partial sentence-strings, it makes these fragments difficult or impossible to translate into certain languages (even just into German or French), as they can have wildly different sentence structures then in English, depending on context.

**Comment by elrido** [comment]
Just using an external library won't solve the XSS problem - that still needs to be solved on our side. 

If we want to solve that for good at it's root cuase, we need to stop using any HTML in our translations. We could no longer use inline links or highlight parts of our strings with em or strong tags. That may be an acceptable tradeoff, if it lets us reduce the attack surface. We could:
- Just accept that we don't highlight strings inside the translations
- Instead of inline links on part of the sentence, just link the whole sentence. Not so elegant, but would avoid having to mix HTML and translated content. It would also simplify the work for translators.

I did review the library that you suggested and see the following challenges with its use for this project:
- They only offer a node client for JavaScript, not a browser library. It would be our responsibility to convert the node module into something that can be used in the browser.
- They offer no PHP library, only one for Python. The whole point of doing our own translations instead of using something existing like GNU gettext (which is supported by PHP) was that I wanted to use the same translation files from both the browser and the server side. Not a blocker, but it would make translators work more difficult, if they had to work with fluent and some other file type.
- Finally they don't address XSS at all, but just offer translated text output, so all of the XSS logic / HTML entity encoding would still have to be implemented by us. So it would neither address the security issue, nor the double encoding ones we try to solve here.

Given all of the above, would you support me (reviewing changes), if I were to attempt a rewrite of our translations to be HTML-free, so we can avoid `.html()` insertions and the need for HTML entity encoding? I could work on it next weekend.

**Comment by rugk** [comment]
I think the current state of this PR is fine, let's merge it. Propbably release a new version.

Yes, a different lib does not solve it, but offloads work, maybe they already handled it. And yeah, also see fluent hardly does so, but I've found https://github.com/projectfluent/fluent/issues/96 and found https://github.com/projectfluent/fluent.js/wiki/DOM-Overlays. This is actually a similar approach that I've used in my previously mentioned project, where I also [keep childrren elements](https://github.com/TinyWebEx/Localizer#keeping-child-elements-in-html) and insert them at the correct position.

> We could:

No, I don't think any of these tradeoffs are good. We _want_ proper links and we _want_ text formatting etc.
So no.

**Comment by elrido** [comment]
Fair enough - I'm ok with the current state as well.

I'll double check the PHP logic (adding a unit test translating that tests a message ID plus parameter with HTML) in a separate PR and get this one merged. I'm in favour of a dot-release with this as well.

**Comment by elrido** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- js/test/I18n.js  13
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PrivateBin/PrivateBin/pullRequest?prid=4811939&bid=16005030)

**Comment by elrido** [comment]
fixed 3 broken cases, but the last test run on travisCI found one more: https://travis-ci.org/PrivateBin/PrivateBin/jobs/644558107#L732 - yes, I did run the unit tests locally 3 times before committing (=> 3 * 100 random strings * 4 cases tested). The property based unit tests are finding edge cases now, as they are designed to do.

**Comment by elrido** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- js/test/I18n.js  1
         

Clones added
============
- js/test/I18n.js  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PrivateBin/PrivateBin/pullRequest?prid=4811939&bid=16005030)

File: js/privatebin.js [hunk 4]
```diff
@@ -630,7 +630,7 @@ jQuery.PrivateBin = (function($, RawDeflate) {
             let containsLinks = args[0].indexOf('<a') !== -1;
 
             // prevent double encoding, when we insert into a text node
-            if (!containsLinks || $element === null) {
+            if (containsLinks || $element === null) {
```
Comment (rugk) [inline js/privatebin.js]: Ugh, last-minute logic changes here? That is core of the XSS prevention.

Comment (elrido) [reply js/privatebin.js]: The unit tests showed cases where the former failed - we had the logic the wrong way around. Glad I added more tests so we don't have to wait for a year and a half to find this. My take-away is: write tests, check they cover all the states in the coverage report and inspect any reported test failure.
